### PR TITLE
Add ignore param to save_hyperparameters

### LINKED ISSUE #6003: Add `ignore` param to `save_hyperparameters`
## 🚀 Enhancement

Add `ignore` param to `self.save_hyperparameters()`. It would support types `list` and `str`.

### Motivation

I believe Users should be given an option to explicitly mention which parameters should not be saved.

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Add `ignore` param to `save_hyperparameters`

Fixes #6003 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1571,7 +1571,7 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
             parents_arguments.update(args)
         return self_arguments, parents_arguments
 
-    def save_hyperparameters(self, *args, frame=None) -> None:
+    def save_hyperparameters(self, *args, ignore: Optional[Union[List, str]] = None, frame=None) -> None:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1574]:
```suggestion
    def save_hyperparameters(self, *args, ignore: Optional[Union[List[str], str]] = None, frame=None) -> None:
```

Can you also annotate frame and add the arguments to the docstring?

File: tests/models/test_hparams.py [hunk 2]
```diff
@@ -695,3 +695,76 @@ def test_hparams(self):
...
+    assert LightningModule.CHECKPOINT_HYPER_PARAMS_KEY in raw_checkpoint
+    assert raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]["test_arg"] == 14
+
+    # verify that model loads correctly
+    model = LocalModel.load_from_checkpoint(raw_checkpoint_path, test_arg2=123)
+    assert model.hparams.test_arg == 14
+    assert "test_arg2" not in model.hparams  # test_arg2 is not registered in class init
+
+
+def test_ignore_args_list_hparams(tmpdir):
```
Comment (carmocca) [inline tests/models/test_hparams.py:736]: Maybe try merging these 2 with pytest parametrize?

Comment (kaushikb11) [reply tests/models/test_hparams.py:736]: Thought about it. But it would get messy/hacky with the assertion of the expected values.

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1615,11 +1621,31 @@ def save_hyperparameters(self, *args, frame=None) -> None:
...
             frame = inspect.currentframe().f_back
         init_args = get_init_args(frame)
         assert init_args, "failed to inspect the self init"
+
+        if ignore:
+            if isinstance(ignore, str):
+                ignore = [ignore]
+            if isinstance(ignore, list):
+                ignore = [arg for arg in ignore if isinstance(arg, str)]
+            init_args = {k: v for k, v in init_args.items() if k not in ignore}
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1647]: how about if `ignore=None` from default?
```
1 not in None <- TypeError: argument of type 'NoneType' is not iterable
```

**Review by Borda — CHANGES_REQUESTED** [review]

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py:1647]: @Borda: I am doing a check for `if ignore`, if it's None, `init_args` won't be changed.

Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py:1647]: And also the other tests will fail, as the default is None :)

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1615,11 +1621,31 @@ def save_hyperparameters(self, *args, frame=None) -> None:
...
+        >>> model.hparams
+        "arg1": 1
+        "arg3": 3.14
         """
         if not frame:
             frame = inspect.currentframe().f_back
         init_args = get_init_args(frame)
         assert init_args, "failed to inspect the self init"
+
+        if ignore:
```
Comment (SkafteNicki) [inline pytorch_lightning/core/lightning.py:1642]:
```suggestion
        if ignore is not None:
```

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1571,12 +1572,17 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
             parents_arguments.update(args)
         return self_arguments, parents_arguments
 
-    def save_hyperparameters(self, *args, frame=None) -> None:
+    def save_hyperparameters(
+        self, *args, ignore: Optional[Union[List[str], str]] = None, frame: Optional[types.FrameType] = None
+    ) -> None:
         """Save all model arguments.
```
Comment (SkafteNicki) [inline pytorch_lightning/core/lightning.py:1578]:
```suggestion
        """Save model arguments to ``hparams`` attribute.
```

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1571,12 +1572,17 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
...
 
         Args:
             args: single object of `dict`, `NameSpace` or `OmegaConf`
-             or string names or arguments from class `__init__`
+                or string names or arguments from class `__init__`
+            ignore: an argument or a list of arguments from class `__init__`
+                to be ignored
+            frame: a frame object. Default is None
```
Comment (SkafteNicki) [inline pytorch_lightning/core/lightning.py:1585]: normally we put examples into an ``Example`` block (everything needs to be indendent then)
```suggestion
            frame: a frame object. Default is None
        
        Example::
```

File: pytorch_lightning/core/lightning.py (hunk 4 above)
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py:1644]: @SkafteNicki Why not just `if ignore`?

Comment (SkafteNicki) [reply pytorch_lightning/core/lightning.py:1643]: it is considered a better practise by pep standards:
https://www.python.org/dev/peps/pep-0008/#programming-recommendations

Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: Updated! :) @SkafteNicki

File: tests/models/test_hparams.py (hunk 2 above)
Comment (carmocca) [reply tests/models/test_hparams.py]: Done in https://github.com/PyTorchLightning/pytorch-lightning/pull/6056/commits/5bd3f6340d257c9afa30c44d095d2001721ba8de

**Review by carmocca — APPROVED** [review]

File: tests/models/test_hparams.py [hunk 7]
```diff
@@ -695,3 +695,39 @@ def test_hparams(self):
         )
         trainer.fit(model)
         _ = TestHydraModel.load_from_checkpoint(checkpoint_callback.best_model_path)
+
+
+@pytest.mark.parametrize("ignore", ("arg2", ("arg2", "arg3")))
+def test_ignore_args_list_hparams(tmpdir, ignore):
```
Comment (kaushikb11) [inline tests/models/test_hparams.py:701]:
```suggestion
def test_ignore_args_hparams(tmpdir, ignore):
```

File: pytorch_lightning/core/lightning.py [hunk 8]
```diff
@@ -1571,13 +1572,19 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
             parents_arguments.update(args)
         return self_arguments, parents_arguments
 
-    def save_hyperparameters(self, *args, frame=None) -> None:
-        """Save all model arguments.
+    def save_hyperparameters(
+        self, *args, ignore: Optional[Union[List[str], str]] = None, frame: Optional[types.FrameType] = None
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1576]:
```suggestion
        self, *args, ignore: Optional[Union[Sequence[str], str]] = None, frame: Optional[types.FrameType] = None
```

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1615,11 +1622,31 @@ def save_hyperparameters(self, *args, frame=None) -> None:
...
         """
         if not frame:
             frame = inspect.currentframe().f_back
         init_args = get_init_args(frame)
         assert init_args, "failed to inspect the self init"
+
+        if ignore is not None:
+            if isinstance(ignore, str):
+                ignore = [ignore]
+            if isinstance(ignore, list):
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1646]: can we have tuples too?

File: pytorch_lightning/core/lightning.py (hunk 8 above)
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: @rohitgr7 Done

File: pytorch_lightning/core/lightning.py [hunk 10]
```diff
@@ -1571,13 +1572,19 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
...
+    def save_hyperparameters(
+        self, *args, ignore: Optional[Union[Sequence[str], str]] = None, frame: Optional[types.FrameType] = None
+    ) -> None:
+        """Save model arguments to ``hparams`` attribute.
 
         Args:
             args: single object of `dict`, `NameSpace` or `OmegaConf`
-             or string names or arguments from class `__init__`
+                or string names or arguments from class `__init__`
+            ignore: an argument or a list of arguments from class `__init__`
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1583]:
```suggestion
                or string names or arguments from class ``__init__``
            ignore: an argument or a list of arguments from class ``__init__``
```

File: pytorch_lightning/core/lightning.py [hunk 11]
```diff
@@ -1571,13 +1572,19 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
...
+                to be ignored
+            frame: a frame object. Default is None
 
+        Example::
         >>> class ManuallyArgsModel(LightningModule):
         ...     def __init__(self, arg1, arg2, arg3):
         ...         super().__init__()
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1590]: I think some indentation is required here to fix the docs.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 11 above)
Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py]: you can have the Example keyword up there. All you need is to indent the examples with a tab to avoid doc failures

Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: Updated!

File: tests/models/test_hparams.py [hunk 12]
```diff
@@ -695,3 +695,39 @@ def test_hparams(self):
...
+    raw_checkpoint_path = _raw_checkpoint_path(trainer)
+    raw_checkpoint = torch.load(raw_checkpoint_path)
+    assert LightningModule.CHECKPOINT_HYPER_PARAMS_KEY in raw_checkpoint
+    assert raw_checkpoint[LightningModule.CHECKPOINT_HYPER_PARAMS_KEY]["arg1"] == 14
+
+    # verify that model loads correctly
+    model = LocalModel.load_from_checkpoint(raw_checkpoint_path, arg2=123, arg3=100)
+    assert model.hparams.arg1 == 14
+    for arg in ignore:
+        assert arg not in model.hparams
```
Comment (awaelchli) [inline tests/models/test_hparams.py:733]: why is training necessary?
testing that hparams are saved to the checkpoint correctly is an independent unit test and does not need to be included here

File: pytorch_lightning/core/lightning.py [hunk 13]
```diff
@@ -1571,55 +1572,84 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
...
+            ...     def __init__(self, arg1, arg2, arg3):
+            ...         super().__init__()
+            ...         # pass argument(s) to ignore as a string or in a list
+            ...         self.save_hyperparameters(ignore='arg2')
+            ...     def forward(self, *args, **kwargs):
+            ...         ...
+            >>> model = ManuallyArgsModel(1, 'abc', 3.14)
+            >>> model.hparams
+            "arg1": 1
+            "arg3": 3.14
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py:1639]: Oh, I only added one with the `ignore` example. Added the `Example` keyword with Identtion.

File: pytorch_lightning/core/lightning.py [hunk 14]
```diff
@@ -1571,55 +1572,84 @@ def _auto_collect_arguments(cls, frame=None) -> Tuple[Dict, Dict]:
...
-        ...     def forward(self, *args, **kwargs):
-        ...         ...
-        >>> model = SingleArgModel(Namespace(p1=1, p2='abc', p3=3.14))
-        >>> model.hparams
-        "p1": 1
-        "p2": abc
-        "p3": 3.14
+                or string names or arguments from class ``__init__``
+            ignore: an argument or a list of arguments from class ``__init__``
+                to be ignored
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py:1587]: Fair point.

File: tests/models/test_hparams.py (hunk 12 above)
Comment (kaushikb11) [reply tests/models/test_hparams.py:733]: Umm, I don't see an issue here. Even if we have a separate test for hparams correctly saved to the checkpoint, it would be the same code/flow defined here.

File: pytorch_lightning/core/lightning.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1639]: yes that's fine given it was already there. It was a general commend because I was surprised the majority of this docstring is about doc-testing all special cases. I would have expected a small concise example like so:

```python
class LitModel(LightningModule):

    def __init__(self, arg0, arg1, arg2=2):
        super().__init__(self)

        # save all parameters
        self.save_hyperparameters()

        # save a selection of parameters
        self.save_hyperparameters("arg0", "arg1")

        # ignore parameters
        self.save_hyperparameters(ignore="arg3")
```

We can improve that later :)

File: pytorch_lightning/core/lightning.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/core/lightning.py]: Updated!

**Comment by kaushikb11** [comment]
@Borda Your requested change.

**Review by tchaton — APPROVED** [review]
LGTM !
